server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -777,6 +994,202 @@ private void insertPendingSegmentIntoMetastore(
+    } else {
+      TimelineObjectHolder<String, DataSegment> existingHolder = Iterables.getOnlyElement(existingChunks);
+      versionOfExistingChunk = existingHolder.getVersion();
+
+      // Don't use the stream API for performance.
+      for (DataSegment segment : FluentIterable
+          .from(existingHolder.getObject())
+          .transform(PartitionChunk::getObject)
+          // Here we check only the segments of the shardSpec which shares the same partition space with the given
+          // partialShardSpec. Note that OverwriteShardSpec doesn't share the partition space with others.
+          // See PartitionIds.
+          .filter(segment -> segment.getShardSpec().sharePartitionSpace(partialShardSpec))) {
+        if (committedMaxId == null
+            || committedMaxId.getShardSpec().getPartitionNum() < segment.getShardSpec().getPartitionNum()) {
+          committedMaxId = SegmentIdWithShardSpec.fromDataSegment(segment);
+        }
+      }
+    }
+
+
+    // Fetch the pending segments for this interval to determine max partitionId
+    // across all shard specs (published + pending).
+    // A pending segment having a higher partitionId must also be considered
+    // to avoid clashes when inserting the pending segment created here.
+    final Set<SegmentIdWithShardSpec> pendingSegments =
+        getPendingSegmentsForIntervalWithHandle(handle, dataSource, interval);
+    if (committedMaxId != null) {
+      pendingSegments.add(committedMaxId);
+    }
+
+    final Map<SegmentCreateRequest, SegmentIdWithShardSpec> createdSegments = new HashMap<>();
+    final Map<String, SegmentIdWithShardSpec> sequenceHashToSegment = new HashMap<>();
+
+    for (SegmentCreateRequest request : requests) {
+      // Check if the required segment has already been created in this batch
+      final String sequenceHash = getSequenceNameAndPrevIdSha(request, interval, skipSegmentLineageCheck);
+
+      final SegmentIdWithShardSpec createdSegment;
+      if (sequenceHashToSegment.containsKey(sequenceHash)) {
+        createdSegment = sequenceHashToSegment.get(sequenceHash);
+      } else {
+        createdSegment = createNewSegment(
+            request,
+            dataSource,
+            interval,
+            versionOfExistingChunk,
+            committedMaxId,
+            pendingSegments
+        );
+
+        // Add to pendingSegments to consider for partitionId
+        if (createdSegment != null) {
+          pendingSegments.add(createdSegment);
+          sequenceHashToSegment.put(sequenceHash, createdSegment);
+          log.info("Created new segment [%s]", createdSegment);
+        }
+      }
+
+      if (createdSegment != null) {
+        createdSegments.put(request, createdSegment);
+      }
+    }
+
+    log.info("Created [%d] new segments for [%d] allocate requests.", sequenceHashToSegment.size(), requests.size());
+    return createdSegments;
+  }
+
+  private SegmentIdWithShardSpec createNewSegment(
```

Comment:
Would it be possible to utilize the older method `createNewSegment` and pass the number of segments to be created as an argument?
If all shardSpecs for common properties can be created with increasing partitionIds in the same call, it may make things simpler and help with performance as well

Reply:
I suppose we could pull out the computation of the `overallMaxId` as that remains the same across the requests in a batch (except including the id of the ids allocated so far). The rest of the stuff in this method will be complicated to pull out because each request can have a different version, depending on the underlying lock that it has created/acquired. Plus, it is not computationally heavy and will not give any perf benefit on commoning out either.

To deduplicate code, the older `createNewSegment` can call this method instead.

Reply:
With time chunk locking, yes. With segment locking, they may be in a batch but still have different versions. This is just because of the way a segment lock is acquired (see `TaskLockbox.tryLock()` and `LockRequestForNewSegment.getVersion()`).  The datasource, groupId and interval are always the same in both the cases. (see `AllocateRequestKey` in `SegmentAllocationQueue`).

The new code is written like this in order to minimize risk by retaining the same behaviour as the old flow except club the metadata calls together.

Reply:
Thanks for the clarification